Batch tx requests

andrewtoth opened the pull request:
Fixes https://github.com/casey/ord/issues/1364 by reverting #1357.

This PR uses `hyper` to fetch raw transactions in the background with batched JSON-RPC requests sent 12 at a time concurrently. This saturates the default 4 rpc threads but doesn't exceed the default rpcworkqueue of 16. I managed to sync to block 776774 in 41 minutes.

This does not require any configuration changes. It uses the json-rpc endpoint, but more efficiently. All missing inputs are sent to a separate thread before indexing, and the resulting input values are sent back via a channel in order.

veryordinally reviewed (commented):
@raphjaph @casey Reviewed conceptually and talked to @andrewtoth - this approach looks consistent with what we discussed yesterday. Does not introduce new communication protocol, and massively speeds up the indexing process for when we have inscriptions (from a day+ to less than an hour). 

Let's discuss and see if you also ACK this approach, then @andrewtoth can complete this PR.

raphjaph commented:
Concept ACK from @veryordinally, @casey, and @raphjaph (reviewing together on a Discord call).

This seems like the right approach. It doesn't require any config changes and doesn't change our security model. 

Would be great to get the MVP in! Should be as simple as possible, with no configuration or additional knobs. Do not worry about polishing too much. Just get full test coverage, undraft, and request review.

andrewtoth commented:
@veryordinally @raphjaph @casey fixed broken test, ready for review.

Also, #1637 and #1731 are relevant to this as well. Since we're using hyper, the json-rpc `index.client` that is created at the start times out and fails when trying to fetch block count again. Same with json-rpc clients created in subcommands before a long update.

raphjaph commented on src/index.rs:
```diff
@@ -20,6 +20,7 @@ use {
 
 mod entry;
 mod rtx;
+mod tx_fetcher;
```
```suggestion
mod fetcher;
```

andrewtoth replied:
I was actually looking at extending this struct to fetch blocks, hashes, headers, and block count. Maybe `HyperJsonRpc` is a better name?

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+  serde::Deserialize,
+};
+
+pub(crate) struct TxFetcher {
```
```suggestion
pub(crate) struct Fetcher {
```

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+  message: String,
+}
+
+impl TxFetcher {
```
```suggestion
impl Fetcher {
```

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+
+    let client = Client::new();
+
+    let url = if url.starts_with("http://") {
```
Let's turn url into a proper URL type.

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+    let (user, password) = auth.get_user_pass()?;
+    let auth = format!("{}:{}", user.unwrap(), password.unwrap());
+    let auth = format!("Basic {}", &base64::encode(auth));
+    Ok(TxFetcher { client, url, auth })
```
```suggestion
    Ok(Fetcher { client, url, auth })
```

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+    let mut reqs = Vec::with_capacity(txids.len());
+    for (i, txid) in txids.iter().enumerate() {
+      let req =
+        format!("{{\"jsonrpc\":\"2.0\",\"id\":\"{i}\",\"method\":\"getrawtransaction\",\"params\":[\"{txid:x}\"]}}");
```
use `json!` macro instead of string

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+
+#[derive(Deserialize, Debug)]
+struct JsonResponse {
+  result: Option<String>,
```
make result a non-opaque type

andrewtoth replied:
It's a string of hex though...

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+struct JsonResponse {
+  result: Option<String>,
+  error: Option<JsonError>,
+  id: String,
```
should this be `String` or `usize`?

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+      ));
+    }
+
+    results.sort_by(|a, b| {
```
why do we need to sort?

andrewtoth replied:
Batched response can be returned in any order, so we need to sort. Our ids are in order so we make sure we send in the same order to the channel.

raphjaph replied:
Thanks for clarifying!

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+        .unwrap()
+        .cmp(&b.id.parse::<usize>().unwrap())
+    });
+
```
add an assert if it's critical that it's sorted

andrewtoth replied:
Sorry, not sure what I would be asserting here? It needs to be sorted before we return, but this is sorting it so we know it's sorted?

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+    Ok(
+      results
+        .into_iter()
+        .map(|res| deserialize(&hex::decode(res.result.unwrap()).unwrap()).unwrap())
```
add comment why `unwraps` are safe

andrewtoth replied:
Some might not necessarily be, if bitcoind is misbehaving. I'll break this out with error checking.

raphjaph commented on src/index/tx_fetcher.rs:
```diff
@@ -0,0 +1,96 @@
+    Ok(
+      results
+        .into_iter()
+        .map(|res| deserialize(&hex::decode(res.result.unwrap()).unwrap()).unwrap())
```
use a more qualified import statement for `deserialize`.

raphjaph commented on src/index/updater.rs:
```diff
@@ -92,6 +98,8 @@ impl Updater {
 
     let rx = Self::fetch_blocks_from(index, self.height, self.index_sats)?;
 
+    let (mut outpoint_sender, mut value_receiver) = Self::spawn_tx_fetcher(index)?;
```
```suggestion
    let (mut outpoint_sender, mut value_receiver) = Self::spawn_fetcher(index)?;
```

raphjaph reviewed (changes requested):
Reviewed by the peanut gallery in Discord: We are confident this fixes the issue but we would love any comments that will help us understand and maintain this code. If you could do that in addition to our review comments (if they are easy), we can smash merge ❤️

andrewtoth commented on Cargo.toml:
```diff
@@ -17,6 +17,7 @@ members = [".", "test-bitcoincore-rpc"]
 anyhow = { version = "1.0.56", features = ["backtrace"] }
 axum = { version = "0.6.1", features = ["headers"] }
 axum-server = "0.4.0"
+base64 = "0.13.1"
```
This is for the rpc basic authentication.

andrewtoth commented:
@raphjaph addresses most of your feedback, asked some questions for clarification on some things.
I added comments to parts of the code that I think could use them. Let me know if there are any other places that could benefit from comments.

kvnn commented:
>  I managed to sync to block 776774 in 41 minutes.
@andrewtoth Nice! Do you mind sharing the machine specs?

andrewtoth commented:
@kvnn running pop_os with Intel® Core™ i7-10870H CPU @ 2.20GHz × 16. Bitcoind was running locally on the same machine.

raphjaph commented on src/index.rs:
```diff
@@ -1026,6 +1027,58 @@ mod tests {
 
+  #[test]
+  fn inscriptions_below_first_inscription_height_are_skipped() {
+    let inscription = inscription("text/plain", "hello");
```
```suggestion
    let inscription = inscription("text/plain;charset=utf-8", "hello");
```

raphjaph reviewed (approved):
Thanks for addressing our comments! Merging this now.
